### PR TITLE
update License and Certificate location

### PR DESCRIPTION
**Pull request by danozdotnet**
update scripts and documentation to use the root directory of the repo for storage of jwt and certificate zip files.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,7 +16,7 @@ This repo should be used in conjunction with the official [SAS Analytics Pro Dep
 1. Follow Step 1 in the `Docker on Linux` section of the official Deployment Guide to make sure that you have your license and certificate files in the top level directory:
    * _SAS-certificates-file-name_.zip
    * _SAS-license-file-name_.jwt
-2. Copy your license `jwt` file into the `sasinside` directory of this repo.
+2. Copy your license `jwt` file into the root directory of this repo.
```
Comment (SelerityMichael) [inline README.md:19]: We can just remove this step from the doco and step 1 already tells users to copy both files to the top level directory.

File: launchapro.sh [hunk 2]
```diff
@@ -45,20 +45,21 @@ if [[ ! -d data ]]; then
   exit 1
 fi
 
-# Get latest license from sasinside directory
-SASLICENSEFILE=$(basename $(ls -r sasinside/*.jwt 2>/dev/null | head -1) 2>/dev/null)
-if [[ "x${SASLICENSEFILE}x" == "xx" ]]; then
-  echo "ERROR: Could not locate SAS license file in sasinside directory"
+# Get latest license file
+SASLICENSEFILE=$(find ~+ -type f -iname "*.jwt" 2>/dev/null | head -1)
```
Comment (SelerityMichael) [inline launchapro.sh:49]: Does this command find the _latest_ file if there are multiple `jwt` files in the directory?  The original `ls -r` command did this by sort all files in reverse order and picking the first one.

File: launchapro.sh [hunk 3]
```diff
@@ -45,20 +45,21 @@ if [[ ! -d data ]]; then
...
+  echo "ERROR: Could not locate SAS license file"
   exit 1
 fi
 export SASLICENSEFILE
 
 # Get latest certificate ZIP
-SASCERTFILE=$(basename $(ls -r ./SASViyaV4_*_certs.zip 2>/dev/null | head -1) 2>/dev/null)
-if [[ "x${SASCERTFILE}x" == "xx" ]]; then
-  echo "ERROR: Could not locate SAS certificate file in current directory"
+SASCERTFILE=$(find ~+ -type f -iname "SASViyaV4_*_certs.zip" 2>/dev/null | head -1)
```
Comment (SelerityMichael) [inline launchapro.sh:57]: Does this command find the _latest_ file if there are multiple `cert` files in the directory?  The original `ls -r` command did this by sort all files in reverse order and picking the first one.

File: README.md (hunk 1 above)
Comment (danozdotnet) [reply README.md]: Yeah I didn't even see that, fixed!

File: launchapro.sh (hunk 2 above)
Comment (danozdotnet) [reply launchapro.sh]: It does now :)

File: launchapro.sh (hunk 3 above)
Comment (danozdotnet) [reply launchapro.sh]: It does now :)

**Review by SelerityMichael — COMMENTED** [review]
Running that `find` command on MacOS fails with:
```
find: -printf: unknown primary or operator
```

I also can't see anywhere that the license file is copied into the `sasinside` directory once it is found?

File: launchapro.sh [hunk 4]
```diff
@@ -45,25 +48,32 @@ if [[ ! -d data ]]; then
...
 fi
 
-# Get latest license from sasinside directory
-SASLICENSEFILE=$(basename $(ls -r sasinside/*.jwt 2>/dev/null | head -1) 2>/dev/null)
-if [[ "x${SASLICENSEFILE}x" == "xx" ]]; then
-  echo "ERROR: Could not locate SAS license file in sasinside directory"
+# Get latest license file, then move to sasinside/
+LINUX_LICENSEFILE=$(find ~+ -type f -iname "*.jwt" -printf '%T@ %p\n' >/dev/null | sort -nr | awk 'NR<=1 {$1=""; print}')
+DARWIN_LICENSEFILE=$(find ~+ -type f -iname "*.jwt" -print0 | xargs -0 stat -f "%m %N" >/dev/null | sort -nr | awk 'NR<=1 {$1=""; print}')
+LICENSEFILE="${HOST_OS^^}_LICENSEFILE"
```
Comment (SelerityMichael) [inline launchapro.sh:54]: The search process should restrict itself to the top-level directory.  If the license file is found it should be _copied_ into the `sasinside` directory.  Current code also finds the license within the sasinside directory.

We should also tighten up the license filename pattern - we know that it should look like `SASViya4_*_license_*.jwt`

File: launchapro.sh [hunk 5]
```diff
@@ -45,25 +48,32 @@ if [[ ! -d data ]]; then
...
+if [[ -n "${!LICENSEFILE}" ]]; then
+  mv "${!LICENSEFILE}" "sasinside/${!LICENSEFILE##*/}"
+  SASLICENSEFILE="sasinside/${!LICENSEFILE##*/}"
```
Comment (SelerityMichael) [inline launchapro.sh:57]: The `SASLICENSEFILE` environment variable should contain the license _filename_ only.  No path.

Comment (danozdotnet) [reply launchapro.sh:57]: ok no problem!

File: launchapro.sh [hunk 6]
```diff
@@ -45,25 +48,32 @@ if [[ ! -d data ]]; then
...
   exit 1
 fi
+
 export SASLICENSEFILE
 
 # Get latest certificate ZIP
-SASCERTFILE=$(basename $(ls -r ./SASViyaV4_*_certs.zip 2>/dev/null | head -1) 2>/dev/null)
-if [[ "x${SASCERTFILE}x" == "xx" ]]; then
-  echo "ERROR: Could not locate SAS certificate file in current directory"
+SASCERTFILE=$(find ~+ -type f -iname "SASViyaV4_*_certs.zip" -printf '%T@ %p\n' | sort -nr | awk 'NR<=1 {print $2}')
```
Comment (SelerityMichael) [inline launchapro.sh:66]: This `find` command needs to be update to work on MacOS as well.

File: launchapro.sh (hunk 4 above)
Comment (danozdotnet) [reply launchapro.sh]: another thought would be to check if a license already exists inside sasinside and whether or not it's newer or older than any others found? or does the docker image work out which license to use itself?

File: launchapro.sh (hunk 6 above)
Comment (danozdotnet) [reply launchapro.sh]: all done now :)

**Review by SelerityMichael — COMMENTED** [review]
Error when testing on MacOs:
```
michaeldixon@Michaels-MBP sas-analytics-pro % ./launchapro.sh 

find: -printf: unknown primary or operator
./launchapro.sh: line 54: ${HOST_OS^^}_LICENSEFILE: bad substitution
ERROR: Could not locate SAS license file
```

**Review by SelerityMichael — COMMENTED** [review]
Still throwing an error:
```
+ [[ ! -d sasinside ]]
+ [[ ! -d data ]]
++ find /Users/michaeldixon/dev/sas-analytics-pro -maxdepth 1 -type f -iname 'SASViya4_*_license_*.jwt' -printf '%T@ %p\n'
++ sort -nr
++ awk 'NR<=1 {$1=""; print}'
+ LINUX_LICENSEFILE=
++ find /Users/michaeldixon/dev/sas-analytics-pro -maxdepth 1 -type f -iname 'SASViya4_*_license_*.jwt' -print0
++ xargs -0 stat -f '%m %N'
++ sort -nr
++ awk 'NR<=1 {$1=""; print}'
+ DARWIN_LICENSEFILE=
launchapro.sh: line 54: ${HOST_OS^^}_LICENSEFILE: bad substitution
+ [[ -n '' ]]
+ echo 'ERROR: Could not locate SAS license file'
ERROR: Could not locate SAS license file
+ exit 1
```

**Review by SelerityMichael — COMMENTED** [review]
Still can't find license file:
```
++ find /Users/michaeldixon/dev/sas-analytics-pro -maxdepth 1 -type f -iname 'SASViya4_*_license_*.jwt' -printf '%T@ %p\n'
++ sort -nr
++ awk 'NR<=1 {$1=""; print}'
+ LINUX_LICENSEFILE=
++ find /Users/michaeldixon/dev/sas-analytics-pro -maxdepth 1 -type f -iname 'SASViya4_*_license_*.jwt' -print0
++ xargs -0 stat -f '%m %N'
++ sort -nr
++ awk 'NR<=1 {$1=""; print}'
+ DARWIN_LICENSEFILE=
+ LICENSEFILE=DARWIN_LICENSEFILE
+ [[ -n '' ]]
+ echo 'ERROR: Could not locate SAS license file.'
ERROR: Could not locate SAS license file.
+ exit 1
```

**Comment by SelerityMichael** [comment]
@danozdotnet did you apply the same logic/update to the Powershell version of the script?  The Mac/Linux version is working now but I can't push until both scripts are in sync.

**Comment by danozdotnet** [comment]
> 
> 
> Error when testing on MacOs:
> 
> ```
> michaeldixon@Michaels-MBP sas-analytics-pro % ./launchapro.sh 
> 
> find: -printf: unknown primary or operator
> ./launchapro.sh: line 54: ${HOST_OS^^}_LICENSEFILE: bad substitution
> ERROR: Could not locate SAS license file
> ```



> 
> 
> @danozdotnet did you apply the same logic/update to the Powershell version of the script? The Mac/Linux version is working now but I can't push until both scripts are in sync.

Not yet, I'll work on that one now :)

**Review by SelerityMichael — APPROVED** [review]
Tested on Win10.

**Pull request merged**